File: appmonet-sdk/appmonet-sdk-mopub/src/main/java/com/monet/bidder/SdkManager.java
package com.monet.bidder;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.os.Build;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;
import android.os.Process;
import android.webkit.ValueCallback;

import androidx.annotation.Nullable;

import com.monet.bidder.threading.InternalRunnable;
import com.mopub.mobileads.MoPubInterstitial;
import com.mopub.mobileads.MoPubView;
import com.mopub.nativeads.MoPubNative;
import com.mopub.nativeads.RequestParameters;

import java.lang.ref.WeakReference;
import java.util.HashMap;
import java.util.Map;

import static com.monet.bidder.Constants.Configurations.MEDIATION_ENABLED;
import static com.monet.bidder.Constants.MISSING_INIT_ERROR_MESSAGE;

/**
 * Created by jose on 8/28/17.
 */

class SdkManager extends BaseManager {
  private final static Logger sLogger = new Logger("SdkManager");
  private static final Object LOCK = new Object();
  private static final int RE_INIT_DELAY = 1000;
  private static SdkManager sInstance;
  private final Map<String, WeakReference<MoPubView>> mopubAdViews = new HashMap<>();
  private final Map<String, AppMonetFloatingAdConfiguration> positions = new HashMap<>();
  static AppMonetConfiguration appMonetConfiguration;
  WeakReference<Activity> currentActivity;

  protected SdkManager(Context context, String applicationId) {
    super(context, applicationId, new MopubAdServerWrapper());
  }

  MoPubView getMopubAdView(String adUnitId) {
    WeakReference<MoPubView> mopubAdView = mopubAdViews.get(adUnitId);
    return (mopubAdView != null) ? mopubAdView.get() : null;
  }

  AppMonetFloatingAdConfiguration getFloatingAdPosition(String adUnitId) {
    return positions.get(adUnitId);
  }

  static void initializeThreaded(final Context context,
      final AppMonetConfiguration appMonetConfiguration) {
    HandlerThread handlerThread =
        new HandlerThread("monet-handler", Process.THREAD_PRIORITY_BACKGROUND);
    handlerThread.start();
    Looper looper = handlerThread.getLooper();
    Handler handler = new Handler(looper);
    handler.post(new InternalRunnable() {
      @Override
      public void runInternal() {
        SdkManager.initialize(context,
            appMonetConfiguration == null ? new AppMonetConfiguration.Builder().build()
                : appMonetConfiguration);
      }

      @Override
      public void catchException(Exception e) {
        sLogger.error("failed to initialize AppMonet SDK: " + e.getLocalizedMessage());
      }
    });
  }

  @SuppressLint("DefaultLocale")
  static void initialize(final Context context, final AppMonetConfiguration appMonetConfiguration) {
    int currentVersion = Build.VERSION.SDK_INT;
    if (currentVersion < Build.VERSION_CODES.JELLY_BEAN) {
      sLogger.warn(
          String.format("Warning! Sdk v%d is not supported. AppMonet SDK Disabled",
              currentVersion));
      return;
    }

    try {
      synchronized (LOCK) {
        if (sInstance != null) {
          sLogger.debug("Sdk has already been initialized. No need to initialize it again.");
          return;
        }
        SdkManager.appMonetConfiguration = appMonetConfiguration;
        sInstance =
            new SdkManager(context.getApplicationContext(), appMonetConfiguration.applicationId);
      }
    } catch (Exception e) {
      if (initRetry < 3) {
        sLogger.error("error initializing ... retrying " + e);
        initRetry += 1;

        Handler handler = new Handler(context.getMainLooper());
        handler.postDelayed(new InternalRunnable() {
          @Override
          public void runInternal() {
            initialize(context, appMonetConfiguration);
          }

          @Override
          public void catchException(Exception e) {
            sLogger.error("Error re-init @ context", e.getMessage());
          }
        }, RE_INIT_DELAY);
      }
    }
  }

  static SdkManager get() {
    return get("");
  }

  @Nullable
  static SdkManager get(String reason) {
    synchronized (LOCK) {
      if (sInstance == null) {
        sLogger.debug(String.format(MISSING_INIT_ERROR_MESSAGE, reason));
      }

      return sInstance;
    }
  }

  MoPubView addBids(MoPubView adView, String adUnitId) {
    logState();

    if (adView == null) {
      sLogger.warn("attempt to add bids to nonexistent AdView");
      return null;
    }

    if (adView.getAdUnitId() == null) {
      sLogger.warn("Mopub adunit id is null. Unable to fetch bids for unit");
      return adView;
    }

    SdkConfigurations config = getSdkConfigurations();
    if (config.getBoolean(MEDIATION_ENABLED)) {
      sLogger.debug("Mediation mode is enabled. Ignoring explicit addBids()");
      return adView;
    }

    MopubAdView mpView = new MopubAdView(adView);
    if (!adUnitId.equals(adView.getAdUnitId())) {
      mpView.setAdUnitId(adUnitId);
    }

    AdServerAdRequest baseRequest;
    registerView(adView, adUnitId);
    AdServerAdRequest request;
    try {
      baseRequest = new MopubAdRequest(adView);
      request = getAuctionManager().addBids(mpView, baseRequest);
    } catch (NullPointerException exp) {
      return null;
    }

    if (request != null) {
      if (request.hasBid()) {
        sLogger.info("found bids for view. attaching");
      } else {
        sLogger.debug("no bids available for request.");
      }

      ((MopubAdRequest) request).applyToView(mpView);
    }
    mopubAdViews.put(mpView.getAdUnitId(), new WeakReference<>(adView));

    return mpView.getMopubView();
  }

  void addBids(final MoPubInterstitial moPubInterstitial, final String alternateAdUnitId,
      final int timeout, final ValueCallback<MoPubInterstitial> onDone) {
    getAuctionManager().timedCallback(timeout, new TimedCallback() {
      @Override
      public void execute(final int remainingTime) {
        if (!isInterstitialActivityRegistered(
            moPubInterstitial.getActivity().getApplicationContext(),
            MonetActivity.class.getName())) {
          String error = "Unable to create activity. Not defined in AndroidManifest.xml. " +
              "Please refer to https://docs.appmonet.com/ for integration infomration.\n";
          sLogger.error(error);
          //todo refactor this.
          getAuctionManager().auctionWebView.trackEvent("integration_error",
              "missing_interstitial_activity", alternateAdUnitId, 0f, 0L);

          throw new ActivityNotFoundException(error);
        }
        logState();
        // todo -> do we need this?
        //        if (appMonetBidder == null) {
        //          onDone.onReceiveValue(moPubInterstitial);
        //          return;
        //        }

        final MopubInterstitialAdView mpView =
            new MopubInterstitialAdView(moPubInterstitial, alternateAdUnitId);
        getAuctionManager().addBids(mpView, new MopubInterstitialAdRequest(moPubInterstitial),
            remainingTime,
            value -> {
              // apply the request to the ad view, and pass that back
              MopubInterstitialAdRequest request = (MopubInterstitialAdRequest) value;
              request.applyToView(mpView);
              onDone.onReceiveValue(moPubInterstitial);
            }
        );
      }

      @Override
      public void timeout() {
        trackTimeoutEvent(alternateAdUnitId, timeout);
        onDone.onReceiveValue(moPubInterstitial);
      }
    });
  }

  void addBids(final MoPubView adView, final String alternateAdUnitId, final int timeout,
      final ValueCallback<MoPubView> onDone) {
    getAuctionManager().timedCallback(timeout, new TimedCallback() {

      @Override
      public void execute(int remainingTime) {
        logState();
        final MopubAdView mpView = new MopubAdView(adView);
        if (adView.getAdUnitId() == null) {
          sLogger.warn("Mopub adunit id is null. Unable to fetch bids for unit");
          onDone.onReceiveValue(adView); // can't continue
          return;
        }

        String adUnit = (alternateAdUnitId == null) ? adView.getAdUnitId() : alternateAdUnitId;
        if (!adView.getAdUnitId().equals(adUnit)) {
          mpView.setAdUnitId(adUnit);
        }

        registerView(adView, adUnit);
        getAuctionManager().addBids(mpView, new MopubAdRequest(adView), remainingTime,
            value -> {
              // apply the request to the ad view, and pass that back
              MopubAdRequest request = (MopubAdRequest) value;
              request.applyToView(mpView);
              mopubAdViews.put(mpView.getAdUnitId(), new WeakReference<>(adView));
              onDone.onReceiveValue(adView);
            }
        );
      }

      @Override
      public void timeout() {
        String adUnit = (alternateAdUnitId == null) ? adView.getAdUnitId() : alternateAdUnitId;
        trackTimeoutEvent(adUnit, timeout);
        onDone.onReceiveValue(adView);
      }
    });
  }

  void addBids(final MoPubNative nativeAd, final RequestParameters requestParameters,
      final String adUnitId,
      final int timeout, final ValueCallback<NativeAddBidsResponse> onDone) {
    getAuctionManager().timedCallback(timeout, new TimedCallback() {
      @Override
      public void execute(final int remainingTime) {
        logState();
        final MopubNativeAdView mpView = new MopubNativeAdView(nativeAd, adUnitId);

        getAuctionManager().addBids(mpView,
            new MopubNativeAdRequest(nativeAd, adUnitId, requestParameters), remainingTime,
            value -> {
              // apply the request to the ad view, and pass that back
              MopubNativeAdRequest request = (MopubNativeAdRequest) value;
              request.applyToView(mpView);
              onDone.onReceiveValue(
                  new NativeAddBidsResponse(nativeAd, request.modifiedRequestParameters));
            }
        );
      }

      @Override
      public void timeout() {
        trackTimeoutEvent(adUnitId, timeout);
        onDone.onReceiveValue(new NativeAddBidsResponse(nativeAd, requestParameters));
      }
    });
  }

  void registerFloatingAd(Activity activity, final AppMonetFloatingAdConfiguration adConfiguration,
      MoPubView moPubView) {
    if (moPubView != null) {
      mopubAdViews.put(adConfiguration.adUnitId, new WeakReference<>(moPubView));
    }
    positions.put(adConfiguration.adUnitId, adConfiguration);
    this.currentActivity = new WeakReference<>(activity);
    getAuctionManager().registerFloatingAd(adConfiguration);
  }

  void unregisterFloatingAd(Activity activity, MoPubView moPubView) {
    if (moPubView != null) {
      mopubAdViews.remove(moPubView.getAdUnitId());
    }
    this.currentActivity = null;
  }

  private void registerView(MoPubView adView, String adUnitIdAlias) {
    if (adView == null) return;

    final String adUnitId = adUnitIdAlias == null ? adView.getAdUnitId() : adUnitIdAlias;
    if (adUnitId == null) {
      sLogger.warn("adView id is null! Cannot register view");
      return;
    }

    final MoPubView.BannerAdListener extant = adView.getBannerAdListener();
    synchronized (LOCK) {
      if (!(extant instanceof MopubBannerAdListener) &&
          !appMonetConfiguration.disableBannerListener && adView.getAdUnitId() != null) {
        sLogger.debug("registering view with internal listener: " + adUnitId);
        adView.setBannerAdListener(new MopubBannerAdListener(adUnitId, extant, this));
      }
    }
  }
}

File: appmonet-sdk/appmonet-sdk-mopub/src/main/java/com/monet/bidder/SdkManager.kt
package com.monet.bidder

import android.annotation.SuppressLint
import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Context
import android.os.Build.VERSION
import android.os.Build.VERSION_CODES
import android.os.Handler
import android.os.HandlerThread
import android.os.Process
import android.webkit.ValueCallback
import com.monet.bidder.AppMonetConfiguration.Builder
import com.monet.bidder.Constants.Configurations.MEDIATION_ENABLED
import com.monet.bidder.Constants.MISSING_INIT_ERROR_MESSAGE
import com.monet.bidder.threading.InternalRunnable
import com.mopub.mobileads.MoPubInterstitial
import com.mopub.mobileads.MoPubView
import com.mopub.nativeads.MoPubNative
import com.mopub.nativeads.RequestParameters
import java.lang.ref.WeakReference
import java.util.HashMap

/**
 * Created by jose on 8/28/17.
 */
internal class SdkManager constructor(
  context: Context?,
  applicationId: String?
) : BaseManager(
    context!!, applicationId, MoPubAdServerWrapper()
) {
  private val mopubAdViews: MutableMap<String, WeakReference<MoPubView>> = HashMap()
  private val positions: MutableMap<String, AppMonetFloatingAdConfiguration> = HashMap()
  var currentActivity: WeakReference<Activity>? = null
  fun getMopubAdView(adUnitId: String?): MoPubView? {
    val mopubAdView = mopubAdViews[adUnitId]
    return mopubAdView?.get()
  }

  fun getFloatingAdPosition(adUnitId: String?): AppMonetFloatingAdConfiguration? {
    return positions[adUnitId]
  }

  fun addBids(
    adView: MoPubView?,
    adUnitId: String
  ): MoPubView? {
    logState()
    if (adView == null) {
      sLogger.warn("attempt to add bids to nonexistent AdView")
      return null
    }
    if (adView.getAdUnitId() == null) {
      sLogger.warn("Mopub adunit id is null. Unable to fetch bids for unit")
      return adView
    }
    val config = sdkConfigurations
    if (config.getBoolean(MEDIATION_ENABLED)) {
      sLogger.debug("Mediation mode is enabled. Ignoring explicit addBids()")
      return adView
    }
    val mpView = MoPubAdView(adView)
    if (adUnitId != adView.getAdUnitId()) {
      mpView.adUnitId = adUnitId
    }
    val baseRequest: AdServerAdRequest
    registerView(adView, adUnitId)
    val request: AdServerAdRequest
    try {
      baseRequest = MoPubAdRequest(adView)
      request = auctionManager.addBids(mpView, baseRequest)
    } catch (exp: NullPointerException) {
      return null
    }
    if (request != null) {
      if (request.hasBid()) {
        sLogger.info("found bids for view. attaching")
      } else {
        sLogger.debug("no bids available for request.")
      }
      (request as MoPubAdRequest).applyToView(mpView)
    }
    mopubAdViews[mpView.adUnitId] = WeakReference(adView)
    return mpView.getMoPubView()
  }

  fun addBids(
    moPubInterstitial: MoPubInterstitial,
    alternateAdUnitId: String?,
    timeout: Int,
    onDone: ValueCallback<MoPubInterstitial?>
  ) {
    auctionManager.timedCallback(timeout, object : TimedCallback {
      override fun execute(remainingTime: Int) {
        if (!isInterstitialActivityRegistered(
                moPubInterstitial.activity.applicationContext,
                MonetActivity::class.java.name
            )
        ) {
          val error = """
                    Unable to create activity. Not defined in AndroidManifest.xml. Please refer to https://docs.appmonet.com/ for integration infomration.

                    """.trimIndent()
          sLogger.error(error)
          //todo refactor this.
          auctionManager.auctionWebView.trackEvent(
              "integration_error",
              "missing_interstitial_activity", alternateAdUnitId!!, 0f, 0L
          )
          throw ActivityNotFoundException(error)
        }
        logState()
        // todo -> do we need this?
        //        if (appMonetBidder == null) {
        //          onDone.onReceiveValue(moPubInterstitial);
        //          return;
        //        }
        val mpView = MoPubInterstitialAdView(moPubInterstitial, alternateAdUnitId!!)
        auctionManager.addBids(mpView, MoPubInterstitialAdRequest(moPubInterstitial),
            remainingTime,
            ValueCallback { value: AdServerAdRequest ->
              // apply the request to the ad view, and pass that back
              val request = value as MoPubInterstitialAdRequest
              request.applyToView(mpView)
              onDone.onReceiveValue(moPubInterstitial)
            }
        )
      }

      override fun timeout() {
        trackTimeoutEvent(alternateAdUnitId, timeout.toFloat())
        onDone.onReceiveValue(moPubInterstitial)
      }
    })
  }

  fun addBids(
    adView: MoPubView,
    alternateAdUnitId: String?,
    timeout: Int,
    onDone: ValueCallback<MoPubView?>
  ) {
    auctionManager.timedCallback(timeout, object : TimedCallback {
      override fun execute(remainingTime: Int) {
        logState()
        val mpView = MoPubAdView(adView)
        if (adView.getAdUnitId() == null) {
          sLogger.warn("Mopub adunit id is null. Unable to fetch bids for unit")
          onDone.onReceiveValue(adView) // can't continue
          return
        }
        val adUnit = alternateAdUnitId ?: adView.getAdUnitId()
        if (adView.getAdUnitId() != adUnit) {
          mpView.adUnitId = adUnit!!
        }
        registerView(adView, adUnit)
        auctionManager.addBids(mpView, MoPubAdRequest(adView), remainingTime,
            ValueCallback { value: AdServerAdRequest ->
              // apply the request to the ad view, and pass that back
              val request = value as MoPubAdRequest
              request.applyToView(mpView)
              mopubAdViews[mpView.adUnitId] = WeakReference(adView)
              onDone.onReceiveValue(adView)
            }
        )
      }

      override fun timeout() {
        val adUnit = alternateAdUnitId ?: adView.getAdUnitId()
        trackTimeoutEvent(adUnit, timeout.toFloat())
        onDone.onReceiveValue(adView)
      }
    })
  }

  fun addBids(
    nativeAd: MoPubNative?,
    requestParameters: RequestParameters?,
    adUnitId: String?,
    timeout: Int,
    onDone: ValueCallback<NativeAddBidsResponse?>
  ) {
    auctionManager.timedCallback(timeout, object : TimedCallback {
      override fun execute(remainingTime: Int) {
        logState()
        val mpView = MopubNativeAdView(nativeAd!!, adUnitId!!)
        auctionManager.addBids(mpView,
            MopubNativeAdRequest(nativeAd, adUnitId, requestParameters), remainingTime,
            ValueCallback { value: AdServerAdRequest ->
              // apply the request to the ad view, and pass that back
              val request = value as MopubNativeAdRequest
              request.applyToView(mpView)
              onDone.onReceiveValue(
                  NativeAddBidsResponse(nativeAd, request.modifiedRequestParameters)
              )
            }
        )
      }

      override fun timeout() {
        trackTimeoutEvent(adUnitId, timeout.toFloat())
        onDone.onReceiveValue(NativeAddBidsResponse(nativeAd!!, requestParameters!!))
      }
    })
  }

  fun registerFloatingAd(
    activity: Activity,
    adConfiguration: AppMonetFloatingAdConfiguration,
    moPubView: MoPubView?
  ) {
    if (moPubView != null) {
      mopubAdViews[adConfiguration.adUnitId] = WeakReference(moPubView)
    }
    positions[adConfiguration.adUnitId] = adConfiguration
    currentActivity = WeakReference(activity)
    auctionManager.registerFloatingAd(adConfiguration.adUnitId, adConfiguration.toJson().toString())
  }

  fun unregisterFloatingAd(
    activity: Activity?,
    moPubView: MoPubView?
  ) {
    if (moPubView != null) {
      mopubAdViews.remove(moPubView.getAdUnitId())
    }
    currentActivity = null
  }

  private fun registerView(
    adView: MoPubView?,
    adUnitIdAlias: String?
  ) {
    if (adView == null) return
    val adUnitId = adUnitIdAlias ?: adView.getAdUnitId()
    if (adUnitId == null) {
      sLogger.warn("adView id is null! Cannot register view")
      return
    }
    val extant = adView.bannerAdListener
    synchronized(LOCK) {
      if (extant !is MopubBannerAdListener &&
          !appMonetConfiguration!!.disableBannerListener && adView.getAdUnitId() != null
      ) {
        sLogger.debug(
            "registering view with internal listener: $adUnitId"
        )
        adView.bannerAdListener = MopubBannerAdListener(adUnitId, extant, this)
      }
    }
  }

  companion object {
    private val sLogger = Logger("SdkManager")
    private val LOCK = Any()
    private const val RE_INIT_DELAY = 1000
    private var sInstance: SdkManager? = null
    var appMonetConfiguration: AppMonetConfiguration? = null
    @JvmStatic fun initializeThreaded(
      context: Context,
      appMonetConfiguration: AppMonetConfiguration?
    ) {
      val handlerThread = HandlerThread("monet-handler", Process.THREAD_PRIORITY_BACKGROUND)
      handlerThread.start()
      val looper = handlerThread.looper
      val handler = Handler(looper)
      handler.post(object : InternalRunnable() {
        override fun runInternal() {
          initialize(
              context,
              appMonetConfiguration ?: Builder().build()
          )
        }

        override fun catchException(e: Exception?) {
          sLogger.error("failed to initialize AppMonet SDK: " + e!!.localizedMessage)
        }
      })
    }

    @SuppressLint("DefaultLocale") fun initialize(
      context: Context,
      appMonetConfiguration: AppMonetConfiguration
    ) {
      val currentVersion = VERSION.SDK_INT
      if (currentVersion < VERSION_CODES.JELLY_BEAN) {
        sLogger.warn(
            String.format(
                "Warning! Sdk v%d is not supported. AppMonet SDK Disabled",
                currentVersion
            )
        )
        return
      }
      try {
        synchronized(LOCK) {
          if (sInstance != null) {
            sLogger.debug("Sdk has already been initialized. No need to initialize it again.")
            return
          }
          Companion.appMonetConfiguration = appMonetConfiguration
          sInstance = SdkManager(context.applicationContext, appMonetConfiguration.applicationId)
        }
      } catch (e: Exception) {
        if (initRetry < 3) {
          sLogger.error(
              "error initializing ... retrying $e"
          )
          initRetry += 1
          val handler = Handler(context.mainLooper)
          handler.postDelayed(object : InternalRunnable() {
            override fun runInternal() {
              initialize(context, appMonetConfiguration)
            }

            override fun catchException(e: Exception?) {
              sLogger.error("Error re-init @ context", e!!.message)
            }
          }, RE_INIT_DELAY.toLong())
        }
      }
    }

    fun get(): SdkManager? {
      return Companion[""]
    }

    @JvmStatic operator fun get(reason: String?): SdkManager? {
      synchronized(LOCK) {
        if (sInstance == null) {
          sLogger.debug(String.format(MISSING_INIT_ERROR_MESSAGE, reason))
        }
        return sInstance
      }
    }
  }
}